[21.05] ceph nautilus refinements

osnyx opened the pull request:
PL-131024

@flyingcircusio/release-managers

## Release process

Impact: internal only

Changelog:
- raise the platform default Ceph release to Nautilus. Jewel and Luminous remain available, but need to be explicitly specified in NixOS config
- cleanups and restructuring of Ceph-related platform code

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - [x] must not introduce new known regressions
  - [x] needs to work on real machines
  - [x] removing the pinning of a ceph release in local config results in Nautilus being used
- [x] Security requirements tested? (EVIDENCE)
  - [x] NixOS tests still pass
  - [x] tested on 1 host in the dev cluster

osnyx committed:
fclib.ceph: restructure mapping of ceph releases per package

- resolves a FIXME
- with more and more packages being dependent on specific ceph releases,
  this is more digestible and keeps the namespace cleaner

osnyx committed:
rgw: supply mime types file

By default, radosgw is looking for mime types specifications under
`/etc/mime.types`, which does not exist.
This fix is not really important, as it is only used by the SWIFT-API
which we do not use at all, but creates a warning at startup that can be
a misleading red herring.

osnyx committed:
ceph.rgw: switch to newer Beast HTTP frontend from Nautilus on

osnyx committed:
ceph: clean up remaining FIXMEs

osnyx committed:
fc-ceph: parameterise maintenance tests to run on multiple release subsystems

osnyx committed:
fc.qemu python3: switch structlog dependency 16.1.0->21.1.0 from nixpkgs

osnyx committed:
fc.qemu-py3: remove unnecessary ceph-16 dependency for our ceph14 (Nautilus) builds

osnyx committed:
ceph.client: use rbd rules from upstream package (Nautilus+)

osnyx committed:
ceph: enable autorepair of detected scrub inconsistencies (Nautilus+)

osnyx committed:
ceph: set default release to Nautilus

ctheune commented on pkgs/fc/ceph/src/fc/ceph/tests/test_maintenance.py:
```diff
@@ -32,9 +32,19 @@ def ceph_json_calls(monkeypatch):
+def test_successful_maintenance_cycle(
+    ceph_json_calls, locktoolcalls, maintenance_manager
+):
+    # only relevant from Nautilus on
```
# only relevant from Nautilus on
     if maintenance_manager.__name__ in ["jewel", "luminous"]:
        return

ctheune commented on pkgs/fc/ceph/src/fc/ceph/tests/test_maintenance.py:
```diff
@@ -59,8 +69,19 @@ def test_successful_maintenance_cycle(ceph_json_calls, locktoolcalls):
+    rbdcalls, locktoolcalls, maintenance_manager
+):
+    # only relevant from Nautilus on
+    if any(
```
see above

ctheune commented on pkgs/fc/ceph/src/fc/ceph/tests/test_maintenance.py:
```diff
@@ -111,8 +143,17 @@ def test_lockimage_created(locktoolcalls, rbdcalls, ceph_json_calls):
 
-def test_tempfail_when_another_lockholder(locktoolcalls):
-    maintenance_task = fc.ceph.maintenance.nautilus.MaintenanceTasks()
+def test_tempfail_when_another_lockholder(locktoolcalls, maintenance_manager):
```
and another option would be to introduce `maintenance_manager_nautilus_plus` oooor with a bit more foresight `maintenance_manager_legacy` and `maintenance_manager` (that only has nautilus+). The latter likely improves maintainability in the future.

osnyx replied:
Done. WHile doing so, I discovered that there currently is no test against non-Nautilus MaintenanceManagers anyways, but I still prepared the `maintenance_manager_legacy` fixture for future use.

ctheune reviewed (changes requested):
Looks good in general, few adjustments.

osnyx committed:
ceph: prevent MANY_OBJECTS_PER_PG from blocking maintenance, raise limit

Currently, one of our clusters is above the "objects per PG" skew limit
in our S3 pools. I (temporarily?) raised the monPgWarnMaxObjectSkew limit until further
consideration.

In general, these kinds of warnings shall not block maintenances.

ctheune commented on pkgs/fc/ceph/src/fc/ceph/tests/conftest.py:
```diff
@@ -0,0 +1,14 @@
+
+
+@pytest.fixture(params=[maintSub.jewel, maintSub.luminous, maintSub.nautilus])
+def maintenance_manager_legacy(request):
```
Uuuuuuhm. Is this actually used anywhere?

osnyx replied:
No. https://github.com/flyingcircusio/fc-nixos/pull/679#discussion_r1206608070
